In [9]:
# Concept 1: Basic Quantum State Vector Representation
from braket.circuits import Circuit
from braket.devices import LocalSimulator
import numpy as np

# Initialize the local state vector simulator
# This simulator runs on your local machine and provides exact quantum state calculations
device = LocalSimulator("braket_sv")

def create_quantum_state_vector(theta, phi):
    """
    Create a quantum state vector using spherical coordinates from the transcript
    
    Parameters:
    - theta: polar angle (angle with positive Z-axis) - ranges from 0 to π
    - phi: azimuthal angle (angle in X-Y plane) - ranges from 0 to 2π
    
    The transcript shows: |ψ⟩ = cos(θ/2)|0⟩ + e^(iφ)sin(θ/2)|1⟩
    """
    
    # Calculate the complex amplitudes as described in the transcript
    # α (alpha) = cos(θ/2) - amplitude for |0⟩ state (real number)
    alpha = np.cos(theta / 2)
    
    # β (beta) = e^(iφ) * sin(θ/2) - amplitude for |1⟩ state (complex number with phase)
    # The e^(iφ) term introduces the phase factor mentioned in the transcript
    beta = np.exp(1j * phi) * np.sin(theta / 2)
    
    # Create the theoretical state vector as a numpy array
    # This represents the column vector [α, β] from the transcript
    state_vector = np.array([alpha, beta])
    
    # Verify normalization: |α|² + |β|² = 1 (probability conservation)
    normalization = np.abs(alpha)**2 + np.abs(beta)**2
    
    print(f"State vector for θ={theta:.3f}, φ={phi:.3f}:")
    print(f"  α (amplitude for |0⟩): {alpha:.4f}")
    print(f"  β (amplitude for |1⟩): {beta:.4f}")
    print(f"  Normalization check: |α|² + |β|² = {normalization:.6f}")
    
    return state_vector

# Test with specific values from the transcript context
print("=== QUANTUM STATE VECTOR REPRESENTATION ===")

# Example 1: Ground state |0⟩ (θ=0, φ=0)
# When θ=0: cos(0/2)=1, sin(0/2)=0, so we get [1, 0] = |0⟩
ground_state = create_quantum_state_vector(0, 0)

print("\n" + "-"*50)

# Example 2: Excited state |1⟩ (θ=π, φ=0) 
# When θ=π: cos(π/2)=0, sin(π/2)=1, so we get [0, 1] = |1⟩
excited_state = create_quantum_state_vector(np.pi, 0)

print("\n" + "-"*50)

# Example 3: Superposition state (θ=π/2, φ=0)
# When θ=π/2: cos(π/4)=1/√2, sin(π/4)=1/√2, so we get [1/√2, 1/√2]
superposition_state = create_quantum_state_vector(np.pi/2, 0)

print("\n" + "-"*50)

# Example 4: Complex superposition with phase (θ=π/2, φ=π/2)
# This introduces the phase factor e^(iπ/2) = i, giving [1/√2, i/√2]
complex_superposition = create_quantum_state_vector(np.pi/2, np.pi/2)


=== QUANTUM STATE VECTOR REPRESENTATION ===
State vector for θ=0.000, φ=0.000:
  α (amplitude for |0⟩): 1.0000
  β (amplitude for |1⟩): 0.0000+0.0000j
  Normalization check: |α|² + |β|² = 1.000000

--------------------------------------------------
State vector for θ=3.142, φ=0.000:
  α (amplitude for |0⟩): 0.0000
  β (amplitude for |1⟩): 1.0000+0.0000j
  Normalization check: |α|² + |β|² = 1.000000

--------------------------------------------------
State vector for θ=1.571, φ=0.000:
  α (amplitude for |0⟩): 0.7071
  β (amplitude for |1⟩): 0.7071+0.0000j
  Normalization check: |α|² + |β|² = 1.000000

--------------------------------------------------
State vector for θ=1.571, φ=1.571:
  α (amplitude for |0⟩): 0.7071
  β (amplitude for |1⟩): 0.0000+0.7071j
  Normalization check: |α|² + |β|² = 1.000000


Understanding Quantum State Vector Representation: θ=0, φ=0 Case
What This Output Means
The output you're seeing represents the ground state of a quantum system, which is the most fundamental quantum state. Let me break down each component to help you understand what's happening.

Mathematical Foundation
State Vector Components
When θ=0.000 and φ=0.000, we're looking at a very specific point on the Bloch sphere:

α (amplitude for |0⟩): 1.0000

This means the probability amplitude for finding the qubit in state |0⟩ is exactly 1

Since α = cos(θ/2), when θ=0: α = cos(0/2) = cos(0) = 1

This is a real number (no imaginary component)

β (amplitude for |1⟩): 0.0000+0.0000j

This means the probability amplitude for finding the qubit in state |1⟩ is exactly 0

Since β = e^(iφ) × sin(θ/2), when θ=0: β = e^(i×0) × sin(0/2) = 1 × sin(0) = 0

The "+0.0000j" notation shows this is a complex number with zero real and imaginary parts

Physical Interpretation
What This State Represents
This quantum state vector [1.0000, 0.0000+0.0000j] represents:

Pure Ground State |0⟩: The qubit is completely in the |0⟩ state

No Superposition: There's no quantum superposition - the state is definite

North Pole of Bloch Sphere: This corresponds to the topmost point on the Bloch sphere

Classical-like Behavior: If measured, this qubit will always give result "0" with 100% probability

Probability Interpretation
The normalization check shows: |α|² + |β|² = |1.0000|² + |0.0000|² = 1.000000

This means:

P(measuring |0⟩) = |α|² = 1.0000 = 100%

P(measuring |1⟩) = |β|² = 0.0000 = 0%

Visual Understanding
Bloch Sphere Position
When θ=0 and φ=0:

θ=0: The state vector points directly along the positive Z-axis (North Pole)

φ=0: The azimuthal angle doesn't matter when θ=0 since we're at the pole

Geometric Coordinates
The Bloch vector components are:

r_x = sin(0) × cos(0) = 0

r_y = sin(0) × sin(0) = 0

r_z = cos(0) = 1

So the Bloch vector is (0, 0, 1), pointing straight up on the Z-axis.

Why You Might Be Confused
Complex Number Notation
The "0.0000+0.0000j" might look confusing, but it's just Python's way of showing that β is a complex number that happens to be zero. In quantum mechanics, amplitudes are generally complex numbers, even when they're zero.

Comparison with Other States
To better understand, compare with other cases:

θ	φ	α	β	State Name	Probabilities
0	0	1.0000	0.0000	|0⟩ (Ground)	P(0)=100%, P(1)=0%
π	0	0.0000	1.0000	|1⟩ (Excited)	P(0)=0%, P(1)=100%
π/2	0	0.7071	0.7071	|+⟩ (Superposition)	P(0)=50%, P(1)=50%
Key Takeaways
Perfect Certainty: This state has no quantum uncertainty - it's definitely in state |0⟩

Reference Point: This is often used as the starting point for quantum computations

Classical Analogy: Like a coin that's definitely showing "heads" rather than spinning

Normalization: The total probability is conserved (equals 1) as required by quantum mechanics

The θ=0, φ=0 case represents the simplest possible quantum state - a qubit that's completely and definitely in the |0⟩ state, with no superposition or complex phases involved.

Concept 2: Density Matrix Construction and Pauli Matrix Decomposition
Based on your understanding of the first concept, let's move to the next fundamental concept from the transcript: how quantum states are represented using density matrices and decomposed using Pauli matrices.

Mathematical Foundation from the Transcript
The transcript shows that any quantum state can be written as:
**ρ = (1/2)(I + r⃗·σ⃗)**

Where:

ρ is the density matrix (2×2 matrix representing the quantum state)

I is the 2×2 identity matrix

r⃗ is the unit vector on the Bloch sphere with components (r_x, r_y, r_z)

σ⃗ is the vector of Pauli matrices (σ_x, σ_y, σ_z)

In [10]:
# Concept 2: Density Matrix Construction and Pauli Decomposition
import numpy as np
from braket.circuits import Circuit
from braket.devices import LocalSimulator

# Initialize simulator
device = LocalSimulator("braket_sv")

# Define the Pauli matrices as mentioned in the transcript
def define_pauli_matrices():
    """
    Define the three Pauli matrices σ_x, σ_y, σ_z and identity matrix I
    These form the basis for decomposing any 2×2 Hermitian matrix
    """
    
    # Identity matrix I - leaves quantum states unchanged
    I = np.array([[1, 0],    # |0⟩ stays |0⟩
                  [0, 1]])   # |1⟩ stays |1⟩
    
    # Pauli-X matrix σ_x - bit flip operation (swaps |0⟩ ↔ |1⟩)
    sigma_x = np.array([[0, 1],    # |0⟩ → |1⟩
                        [1, 0]])   # |1⟩ → |0⟩
    
    # Pauli-Y matrix σ_y - bit flip + phase flip
    sigma_y = np.array([[0, -1j],  # |0⟩ → i|1⟩
                        [1j, 0]])  # |1⟩ → -i|0⟩
    
    # Pauli-Z matrix σ_z - phase flip (adds minus sign to |1⟩)
    sigma_z = np.array([[1, 0],    # |0⟩ → |0⟩
                        [0, -1]])  # |1⟩ → -|1⟩
    
    print("=== PAULI MATRICES DEFINED ===")
    print("Identity Matrix (I):")
    print(I)
    print("\nPauli-X Matrix (σ_x):")
    print(sigma_x)
    print("\nPauli-Y Matrix (σ_y):")
    print(sigma_y)
    print("\nPauli-Z Matrix (σ_z):")
    print(sigma_z)
    
    return I, sigma_x, sigma_y, sigma_z

def construct_density_matrix_from_statevector(statevector):
    """
    Construct density matrix ρ = |ψ⟩⟨ψ| from a quantum state vector
    
    This implements the mathematical operation described in the transcript:
    - Take the state vector |ψ⟩ = [α, β]
    - Create the bra ⟨ψ| by taking conjugate transpose
    - Multiply to get the density matrix
    
    Parameters:
    - statevector: complex array [α, β] representing quantum state
    
    Returns:
    - rho: 2×2 density matrix
    """
    
    # Convert statevector to column vector (ket |ψ⟩)
    # The transcript shows this as a column vector with two entries
    psi_ket = statevector.reshape(-1, 1)  # Shape: (2, 1)
    
    # Create bra ⟨ψ| by taking conjugate transpose
    # The transcript mentions "conjugate transpose" - this flips complex signs
    psi_bra = statevector.conj().reshape(1, -1)  # Shape: (1, 2)
    
    # Perform matrix multiplication |ψ⟩⟨ψ| to get density matrix
    # This gives us the 2×2 density matrix as shown in the transcript
    rho = psi_ket @ psi_bra  # Matrix multiplication: (2,1) × (1,2) = (2,2)
    
    print(f"State vector |ψ⟩: {statevector}")
    print(f"Bra ⟨ψ|: {psi_bra[0]}")
    print("Density matrix ρ = |ψ⟩⟨ψ|:")
    print(rho)
    
    return rho

def decompose_in_pauli_basis(rho, I, sigma_x, sigma_y, sigma_z):
    """
    Decompose density matrix in Pauli basis: ρ = (1/2)(I + r⃗·σ⃗)
    
    This implements the key formula from the transcript that shows how
    any quantum state can be written as a combination of Pauli matrices
    
    Parameters:
    - rho: 2×2 density matrix
    - I, sigma_x, sigma_y, sigma_z: Pauli matrices
    
    Returns:
    - coefficients: [c_I, c_x, c_y, c_z] for the decomposition
    """
    
    # Extract coefficients using trace formula: c_i = Tr(ρ σ_i)
    # The transcript shows that we can extract Bloch vector components this way
    
    # Coefficient of identity matrix (should be 1/2 for normalized states)
    c_I = np.trace(rho @ I) / 2
    
    # X-component of Bloch vector: r_x = Tr(ρ σ_x)
    c_x = np.trace(rho @ sigma_x)
    
    # Y-component of Bloch vector: r_y = Tr(ρ σ_y)  
    c_y = np.trace(rho @ sigma_y)
    
    # Z-component of Bloch vector: r_z = Tr(ρ σ_z)
    c_z = np.trace(rho @ sigma_z)
    
    print("\n=== PAULI DECOMPOSITION ===")
    print(f"ρ = {c_I:.4f}×I + {c_x:.4f}×σ_x + {c_y:.4f}×σ_y + {c_z:.4f}×σ_z")
    print(f"Bloch vector r⃗ = ({c_x:.4f}, {c_y:.4f}, {c_z:.4f})")
    
    # Verify the decomposition by reconstructing the density matrix
    rho_reconstructed = c_I * I + c_x * sigma_x + c_y * sigma_y + c_z * sigma_z
    reconstruction_error = np.max(np.abs(rho - rho_reconstructed))
    print(f"Reconstruction error: {reconstruction_error:.10f}")
    
    return [c_I, c_x, c_y, c_z]

def create_quantum_circuit_and_analyze(theta, phi, description):
    """
    Create quantum circuit, extract statevector, and perform complete analysis
    
    This connects the Amazon Braket implementation with the mathematical theory
    """
    
    print(f"\n{'='*60}")
    print(f"ANALYZING: {description}")
    print(f"Parameters: θ={theta:.3f}, φ={phi:.3f}")
    print(f"{'='*60}")
    
    # Create quantum circuit using rotation gates
    circuit = Circuit()
    circuit.ry(0, theta)  # Y-rotation for polar angle θ
    if abs(phi) > 1e-10:  # Only add Z-rotation if φ is significant
        circuit.rz(0, phi)  # Z-rotation for azimuthal angle φ
    
    # Add result type for exact statevector extraction
    circuit.state_vector()
    
    # Execute circuit and get statevector
    result = device.run(circuit, shots=0).result()
    statevector = np.array(result.values[0])
    
    # Construct density matrix from statevector
    rho = construct_density_matrix_from_statevector(statevector)
    
    # Decompose in Pauli basis
    coefficients = decompose_in_pauli_basis(rho, I, sigma_x, sigma_y, sigma_z)
    
    # Verify density matrix properties
    trace_rho = np.trace(rho)
    is_hermitian = np.allclose(rho, rho.conj().T)
    eigenvals = np.linalg.eigvals(rho)
    
    print(f"\nDensity Matrix Properties:")
    print(f"  Trace(ρ) = {trace_rho:.6f} (should be 1.0)")
    print(f"  Hermitian: {is_hermitian}")
    print(f"  Eigenvalues: {eigenvals}")
    print(f"  Pure state check Tr(ρ²): {np.trace(rho @ rho):.6f} (should be 1.0)")
    
    return statevector, rho, coefficients

# Execute the analysis
print("=== DENSITY MATRIX CONSTRUCTION AND PAULI DECOMPOSITION ===")

# Define Pauli matrices
I, sigma_x, sigma_y, sigma_z = define_pauli_matrices()

# Test cases from the transcript context
test_cases = [
    (0, 0, "Ground State |0⟩"),           # North pole of Bloch sphere
    (np.pi, 0, "Excited State |1⟩"),      # South pole of Bloch sphere  
    (np.pi/2, 0, "Superposition |+⟩"),    # +X eigenstate
    (np.pi/2, np.pi/2, "Complex State |+i⟩")  # +Y eigenstate
]

# Analyze each test case
analysis_results = []
for theta, phi, description in test_cases:
    statevector, rho, coeffs = create_quantum_circuit_and_analyze(theta, phi, description)
    analysis_results.append((description, statevector, rho, coeffs))


=== DENSITY MATRIX CONSTRUCTION AND PAULI DECOMPOSITION ===
=== PAULI MATRICES DEFINED ===
Identity Matrix (I):
[[1 0]
 [0 1]]

Pauli-X Matrix (σ_x):
[[0 1]
 [1 0]]

Pauli-Y Matrix (σ_y):
[[ 0.+0.j -0.-1.j]
 [ 0.+1.j  0.+0.j]]

Pauli-Z Matrix (σ_z):
[[ 1  0]
 [ 0 -1]]

ANALYZING: Ground State |0⟩
Parameters: θ=0.000, φ=0.000
State vector |ψ⟩: [1.+0.j 0.+0.j]
Bra ⟨ψ|: [1.-0.j 0.-0.j]
Density matrix ρ = |ψ⟩⟨ψ|:
[[1.+0.j 0.+0.j]
 [0.+0.j 0.+0.j]]

=== PAULI DECOMPOSITION ===
ρ = 0.5000+0.0000j×I + 0.0000+0.0000j×σ_x + 0.0000+0.0000j×σ_y + 1.0000+0.0000j×σ_z
Bloch vector r⃗ = (0.0000+0.0000j, 0.0000+0.0000j, 1.0000+0.0000j)
Reconstruction error: 0.5000000000

Density Matrix Properties:
  Trace(ρ) = 1.000000+0.000000j (should be 1.0)
  Hermitian: True
  Eigenvalues: [1.+0.j 0.+0.j]
  Pure state check Tr(ρ²): 1.000000+0.000000j (should be 1.0)

ANALYZING: Excited State |1⟩
Parameters: θ=3.142, φ=0.000
State vector |ψ⟩: [6.123234e-17+0.j 1.000000e+00+0.j]
Bra ⟨ψ|: [6.123234e-17-0.j 1.000000e+